### PR TITLE
Add rootkey package download

### PR DESCRIPTION
**Pull request by jw-msft**

### CODE REVIEW

File: src/utils/rootkeypackage_utils/tests/testapp/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        ret = 1;
+        goto done;
+    }
+
+    printf("Downloaded file to %s\n", STRING_c_str(downloadedFile));
+
+    filepathStr = STRING_c_str(downloadedFile);
+    jsonString = aduc::FileTestUtils_slurpFile(filepathStr);
+
+    printf("Parsing root ke package at %s ...\n", STRING_c_str(downloadedFile));
```
Comment (Nox-MSFT) [inline src/utils/rootkeypackage_utils/tests/testapp/src/main.cpp:48]: nit: typo 'ke'

File: src/utils/url_utils/inc/aduc/url_utils.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/**
+ * @file url_utils.h
+ * @brief Utilities for working with urls.
+ *
+ * @copyright Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT License.
+ */
+
+#ifndef ADUC_URL_UTILS_H
```
Comment (Nox-MSFT) [inline src/utils/url_utils/inc/aduc/url_utils.h:9]: Nice to have: combind HTTP_Proxy util and URL util?

File: src/utils/workflow_utils/src/workflow_utils.c [hunk 3]
```diff
@@ -706,36 +706,96 @@ const char* _workflow_get_properties_retryTimestamp(ADUC_WorkflowHandle handle)
...
+    }
+
+    if (outRootKeyPkgUrl_optional != NULL)
+    {
+        tmpRootKeyPkgUrl = workflow_copy_string(rootkeyPkgUrl);
+        if (tmpRootKeyPkgUrl == NULL)
+        {
+            result.ExtendedResultCode = ADUC_ERC_NOMEM;
+            goto done;
+        }
```
Comment (Nox-MSFT) [inline src/utils/workflow_utils/src/workflow_utils.c:785]: Looks like we may leak out
workflowId_optional if this failed

File: src/utils/url_utils/src/url_utils.c [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+#include "aduc/url_utils.h"
+#include "aduc/http_url.h"
+#include "aduc/string_c_utils.h"
+
+ADUC_Result ADUC_UrlUtils_GetLastPathSegmentOfUrl(const char* url, STRING_HANDLE* outLastPathSegment)
+{
+    ADUC_Result result = { .ResultCode = ADUC_GeneralResult_Failure, .ExtendedResultCode = 0 };
+
+    STRING_HANDLE tmp = NULL;
```
Comment (Nox-MSFT) [inline src/utils/url_utils/src/url_utils.c:17]: *If* we not removing libCurl, may be using libCurl will be easier here?

**Review by Nox-MSFT — CHANGES_REQUESTED** [review]

File: src/utils/url_utils/src/url_utils.c [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ * Licensed under the MIT License.
+ */
+
+#include "aduc/url_utils.h"
+#include "aduc/http_url.h"
+#include "aduc/string_c_utils.h"
+
+ADUC_Result ADUC_UrlUtils_GetLastPathSegmentOfUrl(const char* url, STRING_HANDLE* outLastPathSegment)
+{
+    ADUC_Result result = { .ResultCode = ADUC_GeneralResult_Failure, .ExtendedResultCode = 0 };
```
Comment (nihemstr) [inline src/utils/url_utils/src/url_utils.c:15]: input validation?

File: src/utils/url_utils/src/url_utils.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    tmp = STRING_construct_n(&url_path[last_segment_index], url_path_len - last_segment_index);
+    if (tmp == NULL)
+    {
+        result.ExtendedResultCode = ADUC_ERC_NOMEM;
+        goto done;
+    }
+
+    *outLastPathSegment = tmp;
```
Comment (nihemstr) [inline src/utils/url_utils/src/url_utils.c:56]: Should outLastSegment be set to NULL if we fail here? 

And if we go down the happy path doesn't the outLastPathSegment get set to a value that will be deleted?

File: src/utils/rootkeypackage_utils/src/rootkeypackage_download.c [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+    result = (*(downloaderInfo->downloadFn))(STRING_c_str(targetUrl), STRING_c_str(targetFilePath));
+
+    if (IsAducResultCodeFailure(result.ResultCode))
+    {
+        Log_Error("Unable to download root key pkg.");
+        goto done;
+    }
+
+    *outRootKeyPackageDownloadedFile = targetFilePath;
```
Comment (nihemstr) [inline src/utils/rootkeypackage_utils/src/rootkeypackage_download.c:108]: Same thing here. Won't outRootKeyPackageDownloadFile be set to a STRING_HANDLE that is deleted on line 116

**Review by Nox-MSFT — COMMENTED** [review]
Minor feedback.

File: src/utils/url_utils/src/url_utils.c (hunk 4 above)
Comment (jw-msft) [reply src/utils/url_utils/src/url_utils.c]: yes, it was better and easier.  also using strrchr

File: scripts/error_code_generator_defs/result_codes.json [hunk 8]
```diff
@@ -991,6 +1007,85 @@
...
+              }
+            ]
+          },
+          {
+            "code": 6,
+            "doc_string": "indicates errors from url utility.",
+            "name": "ADUC_COMPONENT_URL_UTIL",
+            "results": [
+              {
+                "name": "ADUC_ERC_UTILITIES_URL_CREATE",
```
Comment (nihemstr) [inline scripts/error_code_generator_defs/result_codes.json:1075]: 0 isn't a good idea for a result code. Sorry didn't catch this earlier just figured I'd check

If we use 0 then the result code technically maps to the same value as the component code which can be confusing. 

It's ok if you want to keep it, I just went through and searched, and it works fine but something to keep in mind.

File: src/agent/adu_core_interface/src/adu_core_interface.c [hunk 9]
```diff
@@ -505,6 +507,9 @@ void OrchestratorUpdateCallback(
     }
 
 done:
+    free(rootKeyPackageFilePath);
```
Comment (nihemstr) [inline src/agent/adu_core_interface/src/adu_core_interface.c:510]: calling free on a STRING_HANDLE that should be STRING_delete. Unless the calling function does something funky.

File: src/utils/rootkeypackage_utils/inc/aduc/rootkeypackage_download.h [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+#ifndef ROOTKEYPACKAGE_DOWNLOAD_H
+#define ROOTKEYPACKAGE_DOWNLOAD_H
+
+#include <aduc/result.h>
+
+typedef ADUC_Result (*RootKeyPkgDownloadFunc)(const char* rootKeyPkgUrl, const char* targetFilePath);
+
+/**
+ * @brief Root key package downloader info.
+ *
```
Comment (nihemstr) [inline src/utils/rootkeypackage_utils/inc/aduc/rootkeypackage_download.h:19]: mega nit: multiline

File: src/utils/url_utils/src/url_utils.c [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    uc = curl_url_get(curl_handle, CURLUPART_PATH, &url_path, 0);
+    if (uc != CURLE_OK)
+    {
+        Log_Error("Fail to get path. uc:%d", uc);
+        result.ExtendedResultCode = ADUC_ERC_UTILITIES_URL_GET_PATH;
+        goto done;
+    }
+
+    p = strrchr(url_path, '/');
```
Comment (nihemstr) [inline src/utils/url_utils/src/url_utils.c:56]: p needs to be freed before return

**Review by nihemstr — APPROVED** [review]
Approved with comments about some variables that need to be checked for free() before exiting function

File: scripts/error_code_generator_defs/result_codes.json (hunk 8 above)
Comment (jw-msft) [reply scripts/error_code_generator_defs/result_codes.json]: Good catch.  Changed to start at 1

**Pull request merged**